### PR TITLE
feat: inlay hints typing debounce delay

### LINKED ISSUE #145377: make inlay hints typing debounce max delay configurable
recently, rust-analyzer switched to using vs code's inlay hints provider. however, due to: https://github.com/microsoft/vscode/blob/778b74a2159983ba09568d069ef9c9ecd858f783/src/vs/editor/contrib/inlayHints/browser/inlayHintsController.ts#L214 there is a delay for inlay hints when typing. 

However, I think that this actually makes the experience feel much worse if your language server is able to compute inlay hints quite quickly, and indeed users of the extension have noted this fact as well!

> I have only noticed that after I took the screenshot, I also couldn't reproduce whatever glitch that was. Maybe it's worth noting though, that at least in my case I have noticed that the inlay hints take longer now to render in my VsCode for whatever reason. 

https://github.com/rust-analyzer/rust-analyzer/issues/11733#issuecomment-1069738946

This feature requests proposes to make the hard-coded 1250 be configurable via user settings.

### PR DESCRIPTION
**Pull request by jhgg**
fixes #145377 by adding a configuration called `editor.inlayHints.typingDebounceDelay` - which allows for the configuration of the debounce delay.

### CODE REVIEW

File: src/vs/editor/common/config/editorOptions.ts [hunk 1]
```diff
@@ -2543,6 +2549,13 @@ class EditorInlayHints extends BaseEditorOption<EditorOption.inlayHints, IEditor
 					default: defaults.fontFamily,
 					markdownDescription: nls.localize('inlayHints.fontFamily', "Controls font family of inlay hints in the editor. When set to empty, the `#editor.fontFamily#` is used.")
 				},
+				'editor.inlayHints.typingDebounceDelay': {
+					type: 'number',
+					minimum: 0,
+					maximum: 3000,
+					default: defaults.typingDebounceDelay,
+					markdownDescription: nls.localize('inlayHints.typingDebounceDelay', "Debounce delay for when inlay hints will update after you stop typing.")
```
Comment (gjsjohnmurray) [inline src/vs/editor/common/config/editorOptions.ts:2557]: I suggest you add "in milliseconds" to the description.

**Comment by jhgg** [comment]
Yo, I noticed this was removed from the April 2022 milestone. It would be really really nice if we could include this trivial changeset. It makes the language server/inlay hints feel much more responsive.

**Comment by jhgg** [comment]
Hey @jrieken - would you have a chance to look at this soon? Thank you!

**Comment by Veykril** [comment]
I am not one to usually bump things like this but this issue is still very much a problem. Would it be possible for a team member to take a look at this PR that is now almost 2 years old?

**Comment by davidbarsky** [comment]
Hi all! I'll try summarizing _why_ this issue matters so much for rust-analyzer (and by extension, a large majority of _all_ Rust users who use rust-analyzer via VS Code): the 1250ms debounce delay makes VS Code and rust-analyzer feel _substantially_ more slow than it actually is (for example, rust-analyzer is able to recompute inlay hints in about 2ms on my M1 MacBook Pro, which is about 15 milliseconds faster than my 60Hz monitor is able to refresh!) Having _some_ mechanism—whether that is exposed via LSP capabilities or a setting that the the end-user can set—would make every Rust programmer feel they got a free computer upgrade, courtesy of VS Code.

As a demonstration, I've attached a screen recording of me editing rust-analyzer (at the same commit, using the same rust-analyzer binary)   in both VS Code and Zed. In Zed, I set `"edit_debounce_ms"` to `16`, which roughly corresponds to my monitor's refresh rate.

<details>
<summary>VS Code</summary> 

https://github.com/microsoft/vscode/assets/2067774/6483cbb9-85c0-4acf-9ac0-6144ca437634
</details>


<details>
<summary>Zed</summary> 
https://github.com/microsoft/vscode/assets/2067774/1bfba4cb-e3af-4792-9e2c-ea9bdc994fbe

</details>

I am also aware of Johannes' comments around [flickering](https://github.com/microsoft/vscode/issues/145377#issuecomment-1369501409) and [async updates](https://github.com/microsoft/vscode/issues/145377#issuecomment-1369729604), and while I am sympathetic that a larger design might be necessary, but even changing the default to 100ms would be lovely and provide a day-and-night difference in responsiveness. Granted, I don't know how fast other language servers might be at this, but I can imagine slightly stale inlay hints would be served instead.

The point I'm trying to get at: someone landing this functionality would be _wildly_ impactful, and in my eyes, relatively low effort.

**Comment by jrieken** [comment]
> For context, I work on rust-analyzer full time, and we've determined that a big reason why rust-analyzer feels slow and laggy is the current debouncing behavior in VS Code. [...] Below is a comparison between VS Code and Zed, where we are able to adjust this value:

Thanks for being patient for so long. What Zed has today is how we started, no extra debouncing but direct updates. At the time we had clear feedback, like https://github.com/microsoft/vscode/issues/133730, that inlay hints are "too jumpy", your cursor moves left and right as you type making it hard to follow. 

With https://github.com/microsoft/vscode/issues/145377#issuecomment-1369501409 I have tried to motivate a different solution and that's what I have now implemented in https://github.com/microsoft/vscode/pull/227377. It makes sure inlay hints update as you type but it also reduces the flicker to (very short) typing pauses

**Comment by jrieken** [comment]
Closing this PR without merge because I believe the approach in https://github.com/microsoft/vscode/pull/227377 is best of both worlds. 

You will be able to try this out in next insiders which will ship end of this week.

**Comment by davidbarsky** [comment]
> > For context, I work on rust-analyzer full time, and we've determined that a big reason why rust-analyzer feels slow and laggy is the current debouncing behavior in VS Code. [...] Below is a comparison between VS Code and Zed, where we are able to adjust this value:
> 
> 
> 
> Thanks for being patient for so long. What Zed has today is how we started, no extra debouncing but direct updates. At the time we had clear feedback, like https://github.com/microsoft/vscode/issues/133730, that inlay hints are "too jumpy", your cursor moves left and right as you type making it hard to follow. 
> 
> 
> 
> With https://github.com/microsoft/vscode/issues/145377#issuecomment-1369501409 I have tried to motivate a different solution and that's what I have now implemented in https://github.com/microsoft/vscode/pull/227377. It makes sure inlay hints update as you type but it also reduces the flicker to (very short) typing pauses

Thank you _so_ much for making this happen! I can't wait to try this out in Insiders. For (unlikely) feedback on how it feels, should I open a new issue for discussion?

Regarding the jumpiness: While I'm personally okay with it, I see what you mean and understand why it'd bother some people!

**Comment by jrieken** [comment]
> For (unlikely) feedback on how it feels, should I open a new issue for discussion?

Yeah, new issues are generally preferred and have a higher chance for getting attention
